File: src-tauri/src/models/progress.rs
use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct MediaDestination {
  pub id: String,
  pub group_id: String,
  pub destination: MediaDestinationPath,
  pub is_merged: bool,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct MediaDestinationPath {
  pub confidence: u32,
  pub path: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct MediaProgress {
  pub id: String,
  pub group_id: String,
  pub category: ProgressCategory,
  pub percentage: Option<f64>,
  pub speed_bps: Option<f64>,
  pub eta_secs: Option<u64>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct MediaProgressStage {
  pub id: String,
  pub group_id: String,
  pub stage: ProgressStage,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct MediaProgressComplete {
  pub id: String,
  pub group_id: String,
}

#[derive(Debug, Serialize, Clone)]
pub enum ProgressEvent {
  Destination(MediaDestination),
  Progress(MediaProgress),
  StageChange(MediaProgressStage),
}

#[derive(Debug, Serialize, Clone, PartialEq, Eq, Hash)]
pub enum ProgressCategory {
  Video,
  Audio,
  Subtitles,
  Metadata,
  Thumbnail,
  Other,
}

#[derive(Debug, Serialize, Clone, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum ProgressStage {
  Initializing,
  Downloading,
  Merging,
  Finalizing,
}

File: src/tauri/types/progress.ts
export interface MediaProgressStagePayload {
  id: string;
  groupId: string;
  stage: ProgressStage;
}

export interface MediaProgressPayload {
  id: string;
  groupId: string;
  category: ProgressCategory;
  downloadedBytes?: number;
  totalBytes?: number;
  speedBps?: number;
  etaSecs?: number;
}

export interface MediaProgressPayload {
  id: string;
  groupId: string;
  category: ProgressCategory;
  percentage?: number;
  speedBps?: number;
  etaSecs?: number;
}

export interface MediaGroupProgressPayload extends MediaItemsProgress {
  id: string;
  groupId: string;
  speedBps?: number;
}

export interface MediaItemsProgress {
  total?: number;
  ready?: number;
  done?: number;
  downloading?: number;
}

export interface MediaProgressCompletePayload {
  id: string;
  groupId: string;
}

export enum ProgressCategory {
  video = 'video',
  audio = 'audio',
  subtitles = 'subtitles',
  thumbnail = 'thumbnail',
  metadata = 'metadata',
  other = 'other',
}

export enum ProgressStage {
  initializing = 'initializing',
  downloading = 'downloading',
  merging = 'merging',
  finalizing = 'finalizing',
}

export enum ProgressStyle {
  primary = 'progress-primary',
  warning = 'progress-warning',
  error = 'progress-error',
}
